lib/twom.h

```diff
@@ -0,0 +1,162 @@
+ *   purposes, all without asking permission.
+ */
+
+#ifndef INCLUDED_TWOM_H
+#define INCLUDED_TWOM_H
+
+#include <stdlib.h>
+#include <stdint.h>
+
+struct twom_db;
+struct twom_txn;
+struct twom_cursor;
+
+enum twom_ret {
+    TWOM_OK = 0,
+    TWOM_DONE = 1,
+    TWOM_IOERROR = -1,
+    TWOM_AGAIN = -2,
+    TWOM_EXISTS = -3,
+    TWOM_INTERNAL = -4,
+    TWOM_NOTFOUND = -5,
+    TWOM_LOCKED = -6,
+    TWOM_READONLY = -9,
+};
+
+// we don't reuse flags, there's 32 bits of space, and we spread through them
+enum twom_flagspec {
+    TWOM_CREATE          = 1<<0,    /* Create the database if not existant */
+    TWOM_SHARED          = 1<<1,    /* Open in shared lock mode */
+    TWOM_NOCSUM          = 1<<2,    /* Don't check checksums on read */
+    TWOM_NOSYNC          = 1<<3,    /* Don't msync/fsync on write */
+
+    TWOM_ALWAYSYIELD     = 1<<9,
+    TWOM_NOYIELD         = 1<<10,
+    TWOM_IFNOTEXIST      = 1<<11,
+    TWOM_IFEXIST         = 1<<12,
+    TWOM_FETCHNEXT       = 1<<13,
+    TWOM_SKIPROOT        = 1<<14,
+    TWOM_MVCC            = 1<<15,
+    TWOM_NONBLOCKING     = 1<<16,
+
+    TWOM_CSUM_NULL       = 1<<27,
+    TWOM_CSUM_XXH64      = 1<<28,
+    TWOM_CSUM_EXTERNAL   = 1<<29,
+    TWOM_COMPAR_EXTERNAL = 1<<30
+};
+
+#define TWOM_VERSION 1
+
+typedef int twom_cb(void *rock,
+                    const char *key, size_t keylen,
+                    const char *data, size_t datalen);
+typedef int twom_compar(const char *s1, size_t l1, const char *s2, size_t l2);
+typedef uint32_t twom_csum(const char *s, size_t l);
+
+struct twom_open_data {
+    uint32_t flags;
+    int (*compar)(const char *s1, size_t l1, const char *s2, size_t l2);
+    uint32_t (*csum)(const char *s, size_t l);
+    void (*error)(const char *msg, const char *fmt, ...);
+};
+
+typedef struct twom_open_data twom_init;
+#define TWOM_INIT { 0, NULL, NULL, NULL }
+
+// database operations
+extern int twom_fname_open_db(const char *fname, struct twom_open_data *setup,
+                              struct twom_db **dbptr,
+                              struct twom_txn **tidptr);
+extern int twom_db_close(struct twom_db **dbptr);
+
+// non-transactional operations
+extern int twom_db_fetch(struct twom_db *db,
+                         const char *key, size_t keylen,
+                         const char **keyp, size_t *keylenp,
+                         const char **valp, size_t *vallenp,
+                         int flags);
+extern int twom_db_foreach(struct twom_db *db,
+                           const char *prefix, size_t prefixlen,
+                           twom_cb *p, twom_cb *cb, void *rock,
+                           int flags);
+extern int twom_db_store(struct twom_db *db,
+                         const char *key, size_t keylen,
+                         const char *val, size_t vallen,
+                         int flags);
+
+// utility
+extern int twom_db_consistent(struct twom_db *db);
```

Comment:
Is that a boolean? Since C99 there's the `bool` type and we lean towards requiring C11 anyway? Regardless of the return type, `twom_db_is_consistent` would be clearer.

Reply:
yes, _is_ is better, will do.

Reply:
Hmm... - no, it's not a boolean.  It returns either IOERROR or CHECKSUM depending what's wrong

Reply:
I did rename it, but to "check_consistency"